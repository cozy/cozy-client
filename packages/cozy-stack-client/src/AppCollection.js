import { normalizeDoc } from './DocumentCollection'
import { uri } from './utils'

export const APPS_DOCTYPE = 'io.cozy.apps'
export const KONNECTORS_DOCTYPE = 'io.cozy.konnectors'

export const ENDPOINTS = {
  [APPS_DOCTYPE]: 'apps',
  [KONNECTORS_DOCTYPE]: 'konnectors'
}

export const normalizeApp = (app, doctype) => {
  return {
    ...app,
    ...normalizeDoc(app, doctype),
    ...app.attributes
  }
}

/**
 * Implements methods specific to applications: io.cozy.apps or io.cozy.konnectors
 */
class AppCollection {
  /**
   * @param  {string} doctype App doctype (io.cozy.apps or io.cozy.konnectors)
   * @param  {CozyStackClient} stackClient The stackClient instance to use
   */
  constructor(doctype, stackClient) {
    this.doctype = doctype
    this.stackClient = stackClient
  }

  /**
   * Lists all apps, without filters.
   *
   * The returned documents are not paginated by the stack.
   *
   * @return {{data, meta, skip, next}} The JSON API conformant response.
   * @throws {FetchError}
   */
  async all() {
    const path = uri`/${ENDPOINTS[this.doctype]}/`
    const resp = await this.stackClient.fetchJSON('GET', path)
    return {
      data: resp.data.map(app => normalizeApp(app, this.doctype)),
      meta: {
        count: resp.meta.count
      },
      skip: 0,
      next: false
    }
  }

  async find() {
    throw new Error('find() method is not yet implemented')
  }

  async get() {
    throw new Error('get() method is not yet implemented')
  }

  async create() {
    throw new Error('create() method is not available for applications')
  }

  async update() {
    throw new Error('update() method is not available for applications')
  }

  async destroy() {
    throw new Error('destroy() method is not available for applications')
  }
}

export default AppCollection
