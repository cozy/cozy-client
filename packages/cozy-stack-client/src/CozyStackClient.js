import AppToken from './AppToken'
import DocumentCollection from './DocumentCollection'
import FileCollection from './FileCollection'

const normalizeUri = uri => {
  while (uri[uri.length - 1] === '/') {
    uri = uri.slice(0, -1)
  }
  return uri
}

/**
 * Cozy stack HTTP client.
 */
export default class CozyStackClient {
  constructor({ token, uri = '' }) {
    this.uri = normalizeUri(uri)
    this.setCredentials(token)
  }

  /**
   * Creates a {@link DocumentCollection} instance.
   *
   * @param  {String} doctype The collection doctype.
   * @return {DocumentCollection}
   */
  collection(doctype) {
    if (!doctype) {
      throw new Error('CozyStackClient.collection() called without a doctype')
    }
    switch (doctype) {
      case 'io.cozy.files':
        return new FileCollection(doctype, this)
      default:
        return new DocumentCollection(doctype, this)
    }
  }

  /**
   * Fetches JSON in an authorized way.
   *
   * @param  {String} method The HTTP method.
   * @param  {String} path The URI.
   * @param  {Object} body The payload.
   * @param  {Object} options
   * @return {Object}
   * @throws {FetchError}
   */
  async fetch(method, path, body, options = {}) {
    options.method = method
    const headers = (options.headers = options.headers || {})

    headers['Accept'] = 'application/json'

    if (method !== 'GET' && method !== 'HEAD' && body !== undefined) {
      if (headers['Content-Type']) {
        options.body = body
      } else {
        headers['Content-Type'] = 'application/json'
        options.body = JSON.stringify(body)
      }
    }

    const credentials = options.credentials || this.getCredentials()
    if (credentials) {
      headers['Authorization'] = credentials
      // the option credentials:include tells fetch to include the cookies in the
      // request even for cross-origin requests
      options.credentials = 'include'
    }

    const resp = await fetch(this.fullpath(path), options)
    const contentType = resp.headers.get('content-type')
    const isJson = contentType && contentType.indexOf('json') >= 0
    const data = await (isJson ? resp.json() : resp.text())
    if (resp.ok) {
      return data
    }
    throw new FetchError(resp, data)
  }

  fullpath(path) {
    return this.uri + path
  }

  getCredentials() {
    return this.token ? this.token.toAuthHeader() : null
  }

  setCredentials(token) {
    this.token = token ? new AppToken(token) : null
  }
}

export class FetchError extends Error {
  constructor(response, reason) {
    super()
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor)
    }
    // WARN We have to hardcode this because babel doesn't play nice when extending Error
    this.name = 'FetchError'
    this.response = response
    this.url = response.url
    this.status = response.status
    this.reason = reason

    Object.defineProperty(this, 'message', {
      value:
        reason.message ||
        (typeof reason === 'string' ? reason : JSON.stringify(reason))
    })
  }
}
