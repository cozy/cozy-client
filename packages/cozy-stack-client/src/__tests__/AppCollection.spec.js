jest.mock('../CozyStackClient')

import CozyStackClient from '../CozyStackClient'
import AppCollection, {
  APPS_DOCTYPE,
  KONNECTORS_DOCTYPE
} from '../AppCollection'
import ALL_APPS_RESPONSE from './fixtures/apps.json'
import ALL_KONNECTORS_RESPONSE from './fixtures/apps.json'

const FIXTURES = {
  ALL: {
    [APPS_DOCTYPE]: ALL_APPS_RESPONSE,
    [KONNECTORS_DOCTYPE]: ALL_KONNECTORS_RESPONSE
  }
}
;[APPS_DOCTYPE, KONNECTORS_DOCTYPE].map(appDoctype => {
  describe(`AppCollection with ${appDoctype}`, () => {
    const client = new CozyStackClient()

    describe('all', () => {
      const collection = new AppCollection(appDoctype, client)

      beforeAll(() => {
        client.fetchJSON.mockReturnValue(
          Promise.resolve(FIXTURES.ALL[appDoctype])
        )
      })

      it('should call the right route', async () => {
        await collection.all()
        expect(client.fetchJSON.mock.calls[0]).toMatchSnapshot()
      })

      it('should return a correct JSON API response', async () => {
        const resp = await collection.all()
        expect(resp).toConformToJSONAPI()
      })

      it('should return normalized documents', async () => {
        const resp = await collection.all()
        expect(resp.data[0]).toHaveDocumentIdentity()
      })
    })

    describe('find', () => {
      it('throw error', async () => {
        const collection = new AppCollection(client)
        expect(collection.find()).rejects.toThrowError(
          'find() method is not yet implemented'
        )
      })
    })

    describe('get', () => {
      it('throw error', async () => {
        const collection = new AppCollection(client)
        expect(collection.get()).rejects.toThrowError(
          'get() method is not yet implemented'
        )
      })
    })

    describe('create', () => {
      const collection = new AppCollection(client)

      it('should throw error', async () => {
        expect(collection.create()).rejects.toThrowError(
          'create() method is not available for applications'
        )
      })
    })

    describe('update', () => {
      const collection = new AppCollection(client)

      it('should throw error', async () => {
        expect(collection.update()).rejects.toThrowError(
          'update() method is not available for applications'
        )
      })
    })

    describe('destroy', () => {
      const collection = new AppCollection(client)

      it('should throw error', async () => {
        expect(collection.destroy()).rejects.toThrowError(
          'destroy() method is not available for applications'
        )
      })
    })
  })
})
