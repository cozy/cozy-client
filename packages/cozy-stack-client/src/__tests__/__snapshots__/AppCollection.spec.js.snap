// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppCollection with io.cozy.apps all should call the right route 1`] = `
Array [
  "GET",
  "/apps/",
]
`;

exports[`AppCollection with io.cozy.konnectors all should call the right route 1`] = `
Array [
  "GET",
  "/konnectors/",
]
`;
